File: android/app/src/main/java/com/noop/garmin/GarminSession.kt
package com.noop.garmin

enum class GarminGfdiMessageId(val code: Int) {
    RESPONSE(5000),
    DEVICE_INFORMATION(5024),
    DEVICE_SETTINGS(5026),
    SYSTEM_EVENT(5030),
    SUPPORTED_FILE_TYPES(5031),
    NOTIFICATION_SUBSCRIPTION(5036),
    PROTOBUF_REQUEST(5043),
    CONFIGURATION(5050),
    CURRENT_TIME_REQUEST(5052),
    AUTH_NEGOTIATION(5101);

    companion object { fun fromCode(code: Int): GarminGfdiMessageId? = entries.firstOrNull { it.code == code } }
}

data class GarminHostIdentity(
    val protocolVersion: Int = 150,
    val productNumber: Int = 0xffff,
    val unitNumber: Long = 0xffff_ffffL,
    val softwareVersion: Int = 8300,
    val maximumPacketSize: Int = 0xffff,
    val bluetoothName: String = "NOOP",
    val manufacturer: String = "NOOP",
    val model: String = "Offline companion",
)

enum class GarminSessionError { MALFORMED_REQUEST, INVALID_TIME, FIELD_TOO_LONG, TOO_MANY_CAPABILITIES }
class GarminSessionException(val reason: GarminSessionError) : IllegalArgumentException(reason.name)

data class GarminSessionReply(val outgoing: List<ByteArray>, val configurationCompleted: Boolean = false)

/** Minimal handshake that never claims an unimplemented phone feature. */
object GarminSessionResponder {
    const val SYNC_READY_EVENT = 8

    fun reply(
        message: GarminGfdiMessage,
        nowUnixSeconds: Long,
        timeZoneOffsetSeconds: Int,
        identity: GarminHostIdentity = GarminHostIdentity(),
        supportedCapabilities: ByteArray = byteArrayOf(),
    ): GarminSessionReply? {
        val kind = GarminGfdiMessageId.fromCode(message.messageId) ?: return null
        return when (kind) {
            GarminGfdiMessageId.DEVICE_INFORMATION -> {
                if (message.payload.size < 2) throw GarminSessionException(GarminSessionError.MALFORMED_REQUEST)
                val incomingProtocol = u16(message.payload, 0)
                val body = ArrayList<Byte>()
                responsePrefix(kind, GarminGfdiStatus.ACKNOWLEDGEMENT).forEach(body::add)
                le16(identity.protocolVersion).forEach(body::add)
                le16(identity.productNumber).forEach(body::add)
                le32(identity.unitNumber).forEach(body::add)
                le16(identity.softwareVersion).forEach(body::add)
                le16(identity.maximumPacketSize).forEach(body::add)
                wireString(identity.bluetoothName).forEach(body::add)
                wireString(identity.manufacturer).forEach(body::add)
                wireString(identity.model).forEach(body::add)
                body.add(if (incomingProtocol / 100 == 1) 1 else 0)
                GarminSessionReply(listOf(GarminGfdi.encode(GarminGfdiMessageId.RESPONSE.code, payload = body.toByteArray())))
            }
            GarminGfdiMessageId.CONFIGURATION -> {
                if (supportedCapabilities.size > 255) {
                    throw GarminSessionException(GarminSessionError.TOO_MANY_CAPABILITIES)
                }
                // Mirroring the watch bitset would falsely advertise every feature the watch supports.
                val config = GarminGfdi.encode(
                    kind.code,
                    payload = byteArrayOf(supportedCapabilities.size.toByte()) + supportedCapabilities,
                )
                val ready = GarminGfdi.encode(
                    GarminGfdiMessageId.SYSTEM_EVENT.code,
                    payload = byteArrayOf(SYNC_READY_EVENT.toByte()),
                )
                GarminSessionReply(listOf(config, ready), configurationCompleted = true)
            }
            GarminGfdiMessageId.AUTH_NEGOTIATION -> GarminSessionReply(
                listOf(
                    GarminGfdi.encode(
                        GarminGfdiMessageId.RESPONSE.code,
                        payload = responsePrefix(kind, GarminGfdiStatus.ACKNOWLEDGEMENT) + byteArrayOf(0, 0, 0, 0, 0),
                    ),
                ),
            )
            GarminGfdiMessageId.CURRENT_TIME_REQUEST -> {
                if (message.payload.size < 4) throw GarminSessionException(GarminSessionError.MALFORMED_REQUEST)
                val garminTime = GarminEpoch.garminSeconds(nowUnixSeconds)
                    ?: throw GarminSessionException(GarminSessionError.INVALID_TIME)
                val payload = responsePrefix(kind, GarminGfdiStatus.ACKNOWLEDGEMENT) +
                    message.payload.copyOfRange(0, 4) + le32(garminTime) +
                    le32(timeZoneOffsetSeconds.toLong() and 0xffff_ffffL) + ByteArray(8)
                GarminSessionReply(listOf(GarminGfdi.encode(GarminGfdiMessageId.RESPONSE.code, payload = payload)))
            }
            GarminGfdiMessageId.NOTIFICATION_SUBSCRIPTION,
            GarminGfdiMessageId.PROTOBUF_REQUEST -> GarminSessionReply(
                listOf(
                    GarminGfdi.encode(
                        GarminGfdiMessageId.RESPONSE.code,
                        payload = responsePrefix(kind, GarminGfdiStatus.UNSUPPORTED),
                    ),
                ),
            )
            GarminGfdiMessageId.RESPONSE,
            GarminGfdiMessageId.DEVICE_SETTINGS,
            GarminGfdiMessageId.SYSTEM_EVENT,
            GarminGfdiMessageId.SUPPORTED_FILE_TYPES -> null
        }
    }

    private fun responsePrefix(id: GarminGfdiMessageId, status: GarminGfdiStatus): ByteArray =
        le16(id.code) + status.code.toByte()

    private fun wireString(value: String): ByteArray {
        val bytes = value.toByteArray(Charsets.UTF_8)
        if (bytes.size > 255) throw GarminSessionException(GarminSessionError.FIELD_TOO_LONG)
        return byteArrayOf(bytes.size.toByte()) + bytes
    }

    private fun le16(value: Int): ByteArray = byteArrayOf(value.toByte(), (value ushr 8).toByte())
    private fun le32(value: Long): ByteArray = byteArrayOf(
        value.toByte(), (value ushr 8).toByte(), (value ushr 16).toByte(), (value ushr 24).toByte(),
    )
    private fun u16(bytes: ByteArray, offset: Int): Int =
        (bytes[offset].toInt() and 0xff) or ((bytes[offset + 1].toInt() and 0xff) shl 8)
}

File: Packages/GarminProtocol/Sources/GarminProtocol/GarminSession.swift
import Foundation

public enum GarminGFDIMessageID: UInt16, Sendable {
    case response = 5000
    case deviceInformation = 5024
    case deviceSettings = 5026
    case systemEvent = 5030
    case supportedFileTypes = 5031
    case notificationSubscription = 5036
    case protobufRequest = 5043
    case configuration = 5050
    case currentTimeRequest = 5052
    case authNegotiation = 5101
}

public struct GarminHostIdentity: Equatable, Sendable {
    public let protocolVersion: UInt16
    public let productNumber: UInt16
    public let unitNumber: UInt32
    public let softwareVersion: UInt16
    public let maximumPacketSize: UInt16
    public let bluetoothName: String
    public let manufacturer: String
    public let model: String

    public init(protocolVersion: UInt16 = 150, productNumber: UInt16 = .max,
                unitNumber: UInt32 = .max, softwareVersion: UInt16 = 8300,
                maximumPacketSize: UInt16 = .max, bluetoothName: String = "NOOP",
                manufacturer: String = "NOOP", model: String = "Offline companion") {
        self.protocolVersion = protocolVersion
        self.productNumber = productNumber
        self.unitNumber = unitNumber
        self.softwareVersion = softwareVersion
        self.maximumPacketSize = maximumPacketSize
        self.bluetoothName = bluetoothName
        self.manufacturer = manufacturer
        self.model = model
    }
}

public enum GarminSessionError: Error, Equatable, Sendable {
    case malformedRequest, invalidTime, fieldTooLong, tooManyCapabilities
}

public struct GarminSessionReply: Equatable, Sendable {
    public let outgoing: [[UInt8]]
    public let configurationCompleted: Bool
    public init(outgoing: [[UInt8]], configurationCompleted: Bool = false) {
        self.outgoing = outgoing
        self.configurationCompleted = configurationCompleted
    }
}

/// Minimal, honest GFDI handshake responder. It advertises only the capabilities explicitly supplied
/// by NOOP and returns UNSUPPORTED for phone features that are not implemented yet.
public enum GarminSessionResponder {
    public static let syncReadyEvent: UInt8 = 8

    public static func reply(
        to message: GarminGFDIMessage,
        nowUnixSeconds: Int64,
        timeZoneOffsetSeconds: Int32,
        identity: GarminHostIdentity = .init(),
        supportedCapabilities: [UInt8] = []
    ) throws -> GarminSessionReply? {
        guard let kind = GarminGFDIMessageID(rawValue: message.messageID) else { return nil }
        switch kind {
        case .deviceInformation:
            guard message.payload.count >= 2 else { throw GarminSessionError.malformedRequest }
            let incomingProtocol = u16(message.payload, 0)
            var body = responsePrefix(for: kind, status: .acknowledgement)
            body += le(identity.protocolVersion)
            body += le(identity.productNumber)
            body += le(identity.unitNumber)
            body += le(identity.softwareVersion)
            body += le(identity.maximumPacketSize)
            body += try string(identity.bluetoothName)
            body += try string(identity.manufacturer)
            body += try string(identity.model)
            body.append(incomingProtocol / 100 == 1 ? 1 : 0)
            return .init(outgoing: [try GarminGFDI.encode(messageID: GarminGFDIMessageID.response.rawValue,
                                                          payload: body)])

        case .configuration:
            guard supportedCapabilities.count <= 255 else { throw GarminSessionError.tooManyCapabilities }
            // Do not mirror the watch's bitset: that would claim phone features NOOP has not implemented.
            let config = try GarminGFDI.encode(
                messageID: kind.rawValue,
                payload: [UInt8(supportedCapabilities.count)] + supportedCapabilities
            )
            let ready = try GarminGFDI.encode(
                messageID: GarminGFDIMessageID.systemEvent.rawValue,
                payload: [syncReadyEvent]
            )
            return .init(outgoing: [config, ready], configurationCompleted: true)

        case .authNegotiation:
            let body = responsePrefix(for: kind, status: .acknowledgement) + [0, 0, 0, 0, 0]
            return .init(outgoing: [try GarminGFDI.encode(messageID: GarminGFDIMessageID.response.rawValue,
                                                          payload: body)])

        case .currentTimeRequest:
            guard message.payload.count >= 4,
                  let garminTime = GarminEpoch.garminSeconds(fromUnixSeconds: nowUnixSeconds)
            else { throw GarminSessionError.invalidTime }
            var body = responsePrefix(for: kind, status: .acknowledgement)
            body += Array(message.payload.prefix(4))
            body += le(garminTime)
            body += le(UInt32(bitPattern: timeZoneOffsetSeconds))
            body += [UInt8](repeating: 0, count: 8) // transition end/start: unknown, not fabricated
            return .init(outgoing: [try GarminGFDI.encode(messageID: GarminGFDIMessageID.response.rawValue,
                                                          payload: body)])

        case .notificationSubscription, .protobufRequest:
            let body = responsePrefix(for: kind, status: .unsupported)
            return .init(outgoing: [try GarminGFDI.encode(messageID: GarminGFDIMessageID.response.rawValue,
                                                          payload: body)])

        case .response, .deviceSettings, .systemEvent, .supportedFileTypes:
            return nil
        }
    }

    private static func responsePrefix(for id: GarminGFDIMessageID,
                                       status: GarminGFDIStatus) -> [UInt8] {
        le(id.rawValue) + [status.rawValue]
    }

    private static func string(_ value: String) throws -> [UInt8] {
        let bytes = Array(value.utf8)
        guard bytes.count <= 255 else { throw GarminSessionError.fieldTooLong }
        return [UInt8(bytes.count)] + bytes
    }

    private static func u16(_ bytes: [UInt8], _ offset: Int) -> UInt16 {
        UInt16(bytes[offset]) | UInt16(bytes[offset + 1]) << 8
    }
    private static func le(_ value: UInt16) -> [UInt8] {
        [UInt8(truncatingIfNeeded: value), UInt8(truncatingIfNeeded: value >> 8)]
    }
    private static func le(_ value: UInt32) -> [UInt8] {
        [UInt8(truncatingIfNeeded: value), UInt8(truncatingIfNeeded: value >> 8),
         UInt8(truncatingIfNeeded: value >> 16), UInt8(truncatingIfNeeded: value >> 24)]
    }
}
